File: Python/programmers/etc/level2/행렬_테두리_회전하기.py
def solution(rows, columns, queries):
    answer = []
    board = [[r*columns+(c+1) for c in range(columns)] for r in range(rows)]
    
    for r1, c1, r2, c2 in queries:
        r1, c1, r2, c2 = r1-1, c1-1, r2-1, c2-1
        stack = []

        for a in range(c1, c2+1):
            stack.append(board[r1][a])
            if len(stack) == 1:
                continue
            else:
                board[r1][a] = stack[-2]
        for b in range(r1+1, r2+1):
            stack.append(board[b][a])
            board[b][a] = stack[-2]
        for c in range(c2-1, c1-1, -1):
            stack.append(board[b][c])
            board[b][c] = stack[-2]
        for d in range(r2-1, r1-1, -1):
            stack.append(board[d][c])
            board[d][c] = stack[-2]
        
        answer.append(min(stack))

    return answer


rows, columns = 6, 6
queries = [[2,2,5,4],[3,3,6,6],[5,1,6,3]]
print(solution(rows, columns, queries)) # [8,10,25]

rows, columns = 3, 3
queries = [[1,1,2,2],[1,2,2,3],[2,1,3,2],[2,2,3,3]]
print(solution(rows, columns, queries)) # [1, 1, 5, 3]

rows, columns = 100, 97
queries = [[1,1,100,97]]
print(solution(rows, columns, queries)) # [1]

File: Java/programmers/etc/level2/행렬테두리회전하기.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 행렬테두리회전하기 {
    public int[] solution(int rows, int columns, int[][] queries) {
        int[] answer = new int[queries.length];
        int[][] board = new int[rows][columns];
        
        for (int r=0; r<rows; r++) {
            for (int c=0; c<columns; c++) {
                board[r][c] = r*columns+(c+1);
            }
        }
        
        for (int i=0; i<queries.length; i++) {
            int[] query = queries[i];
            int r1 = query[0]-1, c1 = query[1]-1, r2 = query[2]-1, c2 = query[3]-1;
            List<Integer> stk = new ArrayList<>();
            
            for (int a=c1; a<=c2; a++) {
                stk.add(board[r1][a]);
                if (stk.size() == 1) continue;
                else board[r1][a] = stk.get(stk.size()-2);
            }
            
            for (int b=r1+1; b<=r2; b++) {
                stk.add(board[b][c2]);
                board[b][c2] = stk.get(stk.size()-2);
            }
            
            for (int c=c2-1; c>=c1; c--) {
                stk.add(board[r2][c]);
                board[r2][c] = stk.get(stk.size()-2);
            }
            
            for (int d=r2-1; d>=r1; d--) {
                stk.add(board[d][c1]);
                board[d][c1] = stk.get(stk.size()-2);
            }
            
            int ret = Integer.MAX_VALUE;
            for (int s : stk) {
                ret = Math.min(ret, s);
            }
            answer[i] = ret;
        }
        
        return answer;
    }
}
